// arrow-admin-api/dist/endpoints/modelflows/findAndModifyAPI.js
'use strict';

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var _require = require('./util'),
    getModelParams = _require.getModelParams;

var findAndModifyAPI = function findAndModifyAPI(model) {
	return {
		version: '1',
		returns: {
			status: '$.status|0'
		},
		start: 'model.findAndModify',
		nodes: {
			'model.findAndModify': {
				type: 'ModelFindAndModify',
				config: {
					model: '' + model.name
				},
				parameters: [{
					name: 'limit',
					value: '$.params.limit|0'
				}, {
					name: 'skip',
					value: '$.params.skip|0'
				}, {
					name: 'where',
					value: '$.params.where|0'
				}, {
					name: 'order',
					value: '$.params.order|0'
				}, {
					name: 'sel',
					value: '$.params.sel|0'
				}, {
					name: 'unsel',
					value: '$.params.unsel|0'
				}, {
					name: 'page',
					value: '$.params.page|0'
				}, {
					name: 'per_page',
					value: '$.params.per_page|0'
				}].concat(_toConsumableArray(getModelParams(model))),
				response: {
					context: {},
					routes: [{
						conditions: [{
							key: '$|0',
							exists: false
						}],
						next: 'response.error.notfound'
					}, {
						next: 'response.success'
					}]
				}
			},
			'response.success': {
				type: 'parametermap',
				config: {},
				parameters: [{
					name: 'status',
					default: 204
				}],
				response: {
					context: {
						'$.status|0': 'status'
					},
					routes: [{
						next: null
					}]
				}
			},
			'response.error.notfound': {
				type: 'parametermap',
				config: {},
				parameters: [{
					name: 'status',
					default: 404
				}],
				response: {
					context: {
						'$.status|0': 'status'
					},
					routes: [{
						next: null
					}]
				}
			}
		}
	};
};

exports = module.exports = findAndModifyAPI;
