// arrow-admin-api/dist/endpoints/modelflows/updateAPI.js
'use strict';

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var _require = require('./util'),
    getModelParams = _require.getModelParams;

var updateAPI = function updateAPI(model) {
	return {
		version: '1',
		returns: {
			status: '$.status|0'
		},
		start: 'model.update',
		nodes: {
			'model.update': {
				type: 'ModelUpdate',
				config: {
					model: '' + model.name
				},
				parameters: [{
					name: 'id',
					value: '$.params.id|0'
				}].concat(_toConsumableArray(getModelParams(model))),
				response: {
					context: {},
					routes: [{
						conditions: [{
							key: '$|0',
							exists: false
						}],
						next: 'response.error.notfound'
					}, {
						next: 'response.success'
					}]
				}
			},
			'response.success': {
				type: 'parametermap',
				config: {},
				parameters: [{
					name: 'status',
					default: 204
				}],
				response: {
					context: {
						'$.status|0': 'status'
					},
					routes: [{
						next: null
					}]
				}
			},
			'response.error.notfound': {
				type: 'parametermap',
				config: {},
				parameters: [{
					name: 'status',
					default: 404
				}],
				response: {
					context: {
						'$.status|0': 'status'
					},
					routes: [{
						next: null
					}]
				}
			}
		}
	};
};

exports = module.exports = updateAPI;
