// arrow-admin-api/dist/endpoints/modelflows/upsertAPI.js
'use strict';

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var _require = require('./util'),
    getModelParams = _require.getModelParams;

var upsertAPI = function upsertAPI(model) {
	return {
		version: '1',
		returns: {
			status: '$.status|0',
			headers: {
				location: '$.location|0'
			}
		},
		start: 'model.upsert',
		nodes: {
			'model.upsert': {
				type: 'ModelUpsert',
				config: {
					model: '' + model.name
				},
				parameters: [{
					name: 'id',
					value: '$.params["' + model.name + '"].id|0'
				}].concat(_toConsumableArray(getModelParams(model))),
				response: {
					context: {
						'$.model|0': 'model'
					},
					routes: [{
						conditions: [{
							key: '$.update|0',
							eq: true
						}],
						next: 'response.success.update'
					}, {
						next: 'response.success.create'
					}]
				}
			},
			'response.success.create': {
				type: 'parametermap',
				config: {},
				parameters: [{
					name: 'data',
					value: '$.model.id|0'
				}, {
					name: 'status',
					default: 201
				}],
				response: {
					context: {
						'$.data|0': 'location',
						'$.status|0': 'status'
					},
					routes: [{
						next: null
					}]
				}
			},
			'response.success.update': {
				type: 'parametermap',
				config: {},
				parameters: [{
					name: 'status',
					default: 204
				}],
				response: {
					context: {
						'$.status|0': 'status'
					},
					routes: [{
						next: null
					}]
				}
			}
		}
	};
};

exports = module.exports = upsertAPI;
