// arrow-admin-ui/config.js
/**
 * This is a temporary config file until we implement this properly.
 */
module.exports = exports = {
	// The timestamp format for log entries.  For options, see:
	// https://github.com/felixge/node-dateformat
	'log.list.timestamp.format': 'ddd, mmm d yyyy, h:MM:ss.l TT',
	'log.details.timestamp.format': 'HH:MM:ss.l',
	// default values for the API and UI prefixes.
	// Used if in modes like build/test where arrow isn't available to get the values.
	prefix: '/console',
	apiprefix: '/adminapi/v1',
	apiDocPrefix: '/apidoc',
	docApiUrl: 'http://docs.appcelerator.com/platform/latest/#!/guide/APIs_and_Endpoints',
	docCodeblockUrl: 'http://docs.appcelerator.com/platform/latest/#!/guide/Nodes_-_Codeblock',
	docConfUrl: 'http://docs.appcelerator.com/platform/latest/#!/guide/Console_Configuration',
	docConnUrl: 'http://docs.appcelerator.com/platform/latest/#!/guide/Connectors',
	docLogUrl: 'http://docs.appcelerator.com/platform/latest/#!/guide/API_Builder_Admin_Console-section-src-49153257_APIBuilderAdminConsole-Logs',
	docModelUrl: 'http://docs.appcelerator.com/platform/latest/#!/Models',
	docModelCompUrl: 'http://docs.appcelerator.com/platform/latest/#!/Models_-_Compositing',
	docUIUrl: 'http://docs.appcelerator.com/platform/latest/#!/guide/API_Builder_Admin_Console'
};
