// arrow-admin-api/dist/codeblocks/Codeblock.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * The metadata for the codeblock.
 * @public
 */
var Codeblock = function () {
	/**
  * Create a Codeblock info.
  *
  * @param {object} codeblock - The values to initialize this Codeblock with.
  */
	function Codeblock(codeblock) {
		_classCallCheck(this, Codeblock);

		this.name = codeblock.name;
		this.description = codeblock.description;
		if (!Codeblock.validate(this)) {
			throw new Error('Invalid codeblock: ' + JSON.stringify(codeblock));
		}
	}

	/**
  * The swagger schema for a Codeblock.
  * @return {object} - The schema for a Codeblock.
  */


	_createClass(Codeblock, null, [{
		key: 'validate',


		/**
   * Validate that object is a Codeblock.
   * @param {Codeblock|object} codeblock - The object to validate.
   * @return {boolean} - true if the object is a valid Codeblock.
   */
		value: function validate(codeblock) {
			var valid = true;
			this.schema.required.forEach(function (prop) {
				if (!codeblock.hasOwnProperty(prop) || codeblock[prop] === undefined) {
					valid = false;
				}
			});

			return valid;
		}
	}, {
		key: 'schema',
		get: function get() {
			return {
				title: 'Codeblock',
				type: 'object',
				properties: {
					name: {
						type: 'string'
					},
					description: {
						type: 'string'
					}
				},
				required: ['name']
			};
		}
	}]);

	return Codeblock;
}();

exports = module.exports = Codeblock;
